modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java

```diff
@@ -0,0 +1,192 @@
+		BundleContext bundleContext, JSONFactory jsonFactory) {
+
+		_bundleContext = bundleContext;
+		_jsonFactory = jsonFactory;
+	}
+
+	@Override
+	public Map<String, CSSVariableDescription> addingService(
+		ServiceReference<ServletContext> serviceReference) {
+
+		Map<String, CSSVariableDescription> cssVariablesDescriptions =
+			new ConcurrentHashMap<>();
+
+		modifiedService(serviceReference, cssVariablesDescriptions);
+
+		return cssVariablesDescriptions;
+	}
+
+	@Override
+	public void modifiedService(
+		ServiceReference<ServletContext> serviceReference,
+		Map<String, CSSVariableDescription> cssVariablesDescriptions) {
+
+		ServletContext servletContext = _bundleContext.getService(
+			serviceReference);
+
+		try {
+			InputStream is = servletContext.getResourceAsStream(
+				"/WEB-INF/css-variables.json");
+
+			if (is == null) {
+				return;
+			}
+
+			try {
+				cssVariablesDescriptions.clear();
+
+				cssVariablesDescriptions.putAll(_parse(StringUtil.read(is)));
+			}
+			catch (IllegalArgumentException illegalArgumentException) {
+				_log.error(
+					StringBundler.concat(
+						"Unable to parse css-variables.json of servlet ",
+						"context ", servletContext.getServletContextName()),
+					illegalArgumentException);
+			}
+			catch (IOException ioException) {
+				_log.error(
+					"Unable to read css-variables.json of servlet context" +
+						servletContext.getServletContextName(),
+					ioException);
+			}
+			catch (JSONException jsonException) {
+				_log.error(
+					"Unable to parse css-variables.json of servlet context" +
+						servletContext.getServletContextName(),
+					jsonException);
+			}
+		}
+		finally {
+			_bundleContext.ungetService(serviceReference);
+		}
+	}
+
+	@Override
+	public void removedService(
+		ServiceReference<ServletContext> serviceReference,
+		Map<String, CSSVariableDescription> cssVariableDescriptions) {
+	}
+
+	private CSSVariableType _getCSSVariableType(JSONObject jsonObject) {
+		String type = jsonObject.getString("type");
+
+		if (type.equals("color")) {
+			return CSSVariableType.COLOR;
+		}
+
+		return CSSVariableType.STRING;
+	}
+
+	private Map<String, String> _getLabelsMap(
+		JSONObject cssVariableDefinitionJSONObject, String defaultLabel) {
+
+		Map<String, String> labelsMap = new HashMap<>();
+
+		JSONObject labelsMapJSONObject =
+			cssVariableDefinitionJSONObject.getJSONObject("label");
+
+		if (labelsMapJSONObject != null) {
+			for (String localeKey : labelsMapJSONObject.keySet()) {
+				labelsMap.put(
+					localeKey, labelsMapJSONObject.getString(localeKey));
+			}
+		}
+		else {
+			String label = cssVariableDefinitionJSONObject.getString("label");
+
+			if (label == null) {
+				label = defaultLabel;
+			}
+
+			labelsMap.put(StringPool.BLANK, label);
+		}
+
+		return labelsMap;
+	}
+
+	private Map<String, CSSVariableDescription> _parse(String json)
+		throws JSONException {
+
+		JSONObject jsonObject = _jsonFactory.createJSONObject(json);
```

Comment:
Could you please define the JSON object model somewhere? Feels like an important part of this that is not reflected anywhere.

Reply:
Maybe we can create a JSON schema file, but I'm not sure where to put it (portal, Themes Tookit, ...) :thinking:

Reply:
I see it in https://github.com/liferay-frontend/liferay-portal/pull/17/commits/573a9b32b9d010f06cc8410b8ecdc0a733ccc293#diff-890f20d666168a89aec122633df87dcc, so I'll comment there.